### PR TITLE
hardening/495_configurable_csv_field_separator

### PR DESCRIPTION
**Pull request by pcoello25**
* Implements issue https://github.com/telefonicaid/fiware-cygnus/issues/495
* 100% unit test passed
```
Results :
Tests run: 72, Failures: 0, Errors: 0, Skipped: 0
```

I add some examples that I tested with new code: 

##### CSVROW
Testing with value `csv_separator = "++"` 
```
16/01/12 09:23:23 INFO sinks.OrionHDFSSink: 123456++1970-01-02T10:17:36.789Z++cars++car1++car++speed++float++112.9++hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```
Testing with value `csv_separator = "_"`
```
16/01/12 09:21:35 INFO sinks.OrionHDFSSink: 123456_1970-01-02T10:17:36.789Z_cars_car1_car_speed_float_112.9_hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```
Testing with default value `csv_separator = ","`
```
16/01/12 09:17:41 INFO sinks.OrionHDFSSink: 123456,1970-01-02T10:17:36.789Z,cars,car1,car,speed,float,112.9,hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```

##### CSVCOLUMN
Testing with value `csv_separator = "_"`
```
16/01/12 09:21:35 INFO sinks.OrionHDFSSink: 1970-01-02T10:17:36.789Z_cars_car1_car_112.9_hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```
Testing with value `csv_separator = "@@"`
```
16/01/12 09:24:50 INFO sinks.OrionHDFSSink: 1970-01-02T10:17:36.789Z@@cars@@car1@@car@@112.9@@hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```
Testing with default value `csv_separator = ","`
```
16/01/12 09:17:41 INFO sinks.OrionHDFSSink: 1970-01-02T10:17:36.789Z,cars,car1,car,112.9,hdfs:///user/user1/vehicles/cars/my_cars_speed_float/my_cars_speed_float.txt
```

The value "," has been selected as default (own decision) but we can choose the character that we want. In OrionHDFSSinkTest (line:68) there are a "defaultCsvSeparator" that can be modified. 

* Assignee @frbattid

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -12,3 +12,4 @@
 - [HARDENING] Document the service_as_namespace OrionHDFSSink parameter (#692)
 - [HARDENING] Add a piece of documentation about multitenancy (#607)
 - [HARDENING] Add a piece of documentation about required dependencies to be manually installed when building and installing without dependencies (#635)
+- [HARDENING] Add a configurable value for csv_separator in OrionHDFSSink (#495)
```
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: csv_separator --> CSV separator

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java [hunk 2]
```diff
@@ -263,6 +264,18 @@ public void configure(Context context) {
                     + "properly work!");
         } // if else
         
+        String defaultCsvSeparator = context.getString("default_csv_separator");
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: You can simplify this code by doing the following:

    String csvSeparator = context.getString("csv_separator", ",");

That means "try getting from the configuration a parameter named 'csv_separator'; if not configured, the default CSV separator will the comma".

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java [hunk 3]
```diff
@@ -575,14 +588,14 @@ public void aggregate(CygnusEvent cygnusEvent) throws Exception {
                 
                 // create a line and aggregate it
                 String line = "{"
-                    + "\"" + Utils.encodeHive(Constants.RECV_TIME_TS) + "\":\"" + recvTimeTs / 1000 + "\","
-                    + "\"" + Utils.encodeHive(Constants.RECV_TIME) + "\":\"" + recvTime + "\","
+                    + "\"" + Utils.encodeHive(Constants.RECV_TIME_TS) + "\":\"" + recvTimeTs / 1000 + "\","
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: Nothing has been modified at these lines :)

File: src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java [hunk 4]
```diff
@@ -466,6 +467,7 @@ private Context createContext(boolean useDeprecatedParams, String fileFormat) {
             context.put("hdfs_username", hdfsUsername);
         } // if else
         
+        context.put("default_csv_separator",defaultCsvSeparator);
```
Comment (frbattid) [inline src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java]: The only configuration parameter must be "csv_separator".

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (pcoello25) [inline CHANGES_NEXT_RELEASE]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/93781aa1f586022ef98d1f1612a7c737add79f30

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java (hunk 2 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/93781aa1f586022ef98d1f1612a7c737add79f30

File: src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java (hunk 4 above)
Comment (pcoello25) [inline src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/93781aa1f586022ef98d1f1612a7c737add79f30

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java [hunk 5]
```diff
@@ -263,6 +264,14 @@ public void configure(Context context) {
                     + "properly work!");
         } // if else
         
+        String csvSeparator = context.getString("csv_separator", ",");        
+        if (csvSeparator != null && csvSeparator.length() > 0) {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: These lines are not necessary sicne there is always a value for "csv_separator".

File: src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java [hunk 6]
```diff
@@ -466,6 +467,7 @@ private Context createContext(boolean useDeprecatedParams, String fileFormat) {
             context.put("hdfs_username", hdfsUsername);
         } // if else
         
+        context.put("csv_separator",csvSeparator);
```
Comment (frbattid) [inline src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java]: A white space is required after the comma.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java (hunk 3 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: This code belongs to json method, but I had to change the csv method. The changes are spaces which doesn't affect to the main code.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java (hunk 5 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSink.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/d89500c46c3c725a627c2b47e0fe6214cef0d6ad

File: src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java (hunk 6 above)
Comment (pcoello25) [inline src/test/java/com/telefonica/iot/cygnus/sinks/OrionHDFSSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/d89500c46c3c725a627c2b47e0fe6214cef0d6ad

**Comment by frbattid** [comment]
LGTM to the code, the documentation is missing.

**Comment by pcoello25** [comment]
Added csv_separator parameter in the documentation in https://github.com/telefonicaid/fiware-cygnus/commit/d196f8fcd2e24ad192b66b13e2fc2ade8328d772

**Comment by frbattid** [comment]
LGTM

**Pull request merged**